### DISCUSSION TITLE
Setting platform and mission specific configuration parameters

### DISCUSSION THREAD

**Discussion by 9a3eedi** [nasa/cFS#734] [Q&A] [answered]
Hello,

While looking to implement system time on a raspberry pi as the timestamp source for messages (based on this discussion https://github.com/nasa/cFS/discussions/446) I've been trying to figure out _the right way_ to set CFE_PLATFORM_TIME_CFG_SOURCE and CFE_PLATFORM_TIME_CFG_SRC_TIME to true.

I did find those settings under default_cfe_time_internal_cfg.h ... but those are supposed to be defaults. After some digging I realized that CMake generates headers that include these defaults, by default, and that I should provide a separate header file somewhere to override these defaults. The old way of doing this is by writing a single cfe_mission_cfg.h or cfe_platform_cfg.h file, somewhere, to override any setting. But this is not recommended anymore, and I should override individual configuration headers manually now.

What's not clear to me is how do I go about this. I need to write a header and override those defines, that much I know, but where should it go, and how does it get included in the build process? Should I include anything in my header? Should I use #undef before using #define?

In general, I've been having trouble understanding the build system used in cFS and how a developer would use it, and I couldn't find any guide walking through how it works apart from READMEs and little bits of information here and there. Is there a guide that I should be following?

**Comment by irowebbn** [comment, marked as answer]
I believe the conventional location is in `<arch>_platform_cfg.h`, which goes in a directory called `<MISSIONCONFIG>_defs` at the same level as cFE in your project tree. For an example, see [`cfe/cmake/sample_defs/example_platform_cfg.h`](https://github.com/nasa/cFE/blob/37f1d28df44aa7a858555b92f0c20310a6803b9f/cmake/sample_defs/example_platform_cfg.h). This folder is typically where you will want to put any of your override headers.

And yes , unfortunately, the build system is not well-documented.  In my opinion, your best bet is the [README under `cfe/cmake`](https://github.com/nasa/cFE/blob/37f1d28df44aa7a858555b92f0c20310a6803b9f/cmake/README.md)  (although this hasn't been updated in some time) and the headers of the CMake files themselves, especially [`arch_build.cmake`](https://github.com/nasa/cFE/blob/37f1d28df44aa7a858555b92f0c20310a6803b9f/cmake/arch_build.cmake).
 
1. Create a `mission_defs` folder. You can get an example at [`cfe/cmake/sample_defs/`](https://github.com/nasa/cFE/tree/37f1d28df44aa7a858555b92f0c20310a6803b9f/cmake/sample_defs). There are comments in some of the files indicating what each one is used for. You can have multiple mission definitions folders, just replace "mission" with any name and set the environment variable `MISSIONCONFIG` to that name before building.
2. Edit the [`misison_defs/targets.cmake`](https://github.com/nasa/cFE/blob/37f1d28df44aa7a858555b92f0c20310a6803b9f/cmake/sample_defs/targets.cmake) file to specify which apps to build and for which architectures. This file has detailed comments explaining how to use it.
3. Configure the build using CMake and build using the build tool (currently only Unix Makefiles, ie `make`, are supported):

```bash
export MISSIONCONFIG=<my_mission> # optional if there is only one mission configuration at `mission_defs`
# Assuming you start from the top-level folder containing cFE and your `apps` folder
mkdir build
cd build
cmake  -S  ../cfe # Configures and generates build system using the CMakeLists.txt in cFE
cmake --build -- mission-all # Can also just run `make mission-all`
cmake --build -- mission-install # Can also just run `make mission-install`
```

Other targets you can use:
`all` : Build the tools that execute on the development host  (this is a misnomer, not really "all")
`mission-doc`: Build the Doxygen for the apps in your system
`cfe-usersguide`: Build the Doxygen for cFE.

To see all the targets, run `cmake --build -- help` or `make help`

You can pass variables to CMake to change how the build is configured. For example,   `-DCMAKE_INSTALL_PREFIX=exe -DENABLE_UNIT_TESTS=true -DCMAKE_BUILD_TYPE=Debug`

You'll see in some of the docs like the cFS README and cFE CMake README that there is a top-level Makefile you can use (copied from cfe/cmake/Makefile.sample). This is really just a wrapper for calling CMake and I personally don't recommend using it since that is a very unconventional way to invoke CMake.

**Reply by irowebbn** [reply]
For a better explanation of how cFS uses CMake, see [this presentation](https://flightsoftware.jhuapl.edu/files/2017/cFS-Day/cFS17-C3-cfe-cmake-build-2017-workshop.pdf)

**Reply by 9a3eedi** [reply]
The presentation is actually very helpful and exactly what I needed, thanks!
